api.go

```diff
@@ -579,7 +588,15 @@ func StatusContainer(podID, containerID string) (ContainerStatus, error) {
 		return ContainerStatus{}, errNeedContainerID
 	}
 
-	lockFile, err := lockPod(podID)
+	var wg sync.WaitGroup
+	// Deferring the synchronization here is very important since we want
+	// to avoid a deadlock. Indeed, the goroutine started by statusContainer
```

Comment:
So the `statusContainer` go routine will have to wait for all read locks to be released before running. That means it will run asynchronously, but it does not mean there will be any deadlock. Am I missing something ?

Running asynchronously is fine, as long as the container status is properly set at some point.

Reply:
There are two things here. First, we have to wait for the end of the goroutine, otherwise the runtime could exit and kill the goroutine when exiting, meaning the go routine could be interrupted while updating the container status.
Then, because we have to wait, we use here a WaitGroup and the call to `wg.Wait()` has to happen after the lock has been released from this function. Because the unlock is called also through a defer function, we have to make sure the `defer wg.Wait()` is earlier in the code so that it is called as the last thing.